Report network download progress

pvollan opened the pull request:
#### c804c90d7236810c046ffdae6ac977bec91f6ad4
<pre>
Report network download progress
<a href="https://rdar.apple.com/129296777">rdar://129296777</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275176">https://bugs.webkit.org/show_bug.cgi?id=275176</a>

Reviewed by Chris Dumez.

Adopt modern download progress reporting mechanism. This is replacing our current use of the
NSProgress API for reporting download progress. The feature flag is not being set in this
patch, so there should be no behavior change from this patch. This patch is also replacing
the use of sandbox SPI to create and consume a sandbox extension to the location of the
downloaded file. This is being replaced with the bookmark API.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebKit/NetworkProcess/Downloads/Download.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
(WebKit::DownloadManager::publishDownloadProgress):
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
(WebKit::PendingDownload::publishProgress):
(WebKit::PendingDownload::didBecomeDownload):
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):
* Source/WebKit/NetworkProcess/Downloads/cocoa/WKDownloadProgress.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/WKDownloadProgress.mm:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::publishDownloadProgress):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
(WebKit::DownloadProxy::publishProgress): Deleted.
* Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm: Added.
(WebKit::DownloadProxy::publishProgress):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/281152@main">https://commits.webkit.org/281152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7eb95b1b5ad144be59d2f89fab0424257ad5d673

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58882 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11373 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62514 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9326 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45861 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9527 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47623 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6646 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60913 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35761 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50930 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28480 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32491 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8330 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54448 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8494 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64215 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2795 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8475 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54946 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2804 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50950 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55043 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2350 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8807 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34040 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35124 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36209 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34870 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
```diff
@@ -75,17 +75,31 @@ void PendingDownload::cancel(CompletionHandler<void(std::span<const uint8_t>)>&&
+{
+    ASSERT(!m_progressURL.isValid());
+    m_progressURL = url;
+    m_bookmarkData = WTFMove(bookmarkData);
```
Is this really safe? `m_bookmarkData` is a span, so we don't own the bytes. What makes sure the memory the span is pointed to in `m_bookmarkData` remains valid between this function call and the `didBecomeDownload` function call which ends up using the bookmark data?

pvollan replied:
That is a great catch! I have changed the `m_bookmarkData` member to a Vector, and addressed further issues.

Thanks for reviewing!

cdumez commented on Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
```diff
@@ -75,17 +75,31 @@ void PendingDownload::cancel(CompletionHandler<void(std::span<const uint8_t>)>&&
 
 void PendingDownload::didBecomeDownload(const std::unique_ptr<Download>& download)
 {
+#if HAVE(MODERN_DOWNLOADPROGRESS)
```
Maybe we can write this like so:
```
if (!m_progressURL.isValid())
    return;
#if HAVE(MODERN_DOWNLOADPROGRESS)
download->publishProgress(m_progressURL, WTFMove(m_bookmarkData));
#else
download->publishProgress(m_progressURL, WTFMove(m_progressSandboxExtension));
#endif
```
to share a little more code.

cdumez commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -85,6 +85,13 @@
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+void Download::publishProgress(const URL& url, std::span<const uint8_t>&& bookmarkData)
+{
+    auto bookmark = adoptNS([[NSData alloc] initWithBytes:bookmarkData.data() length:bookmarkData.size()]);
```
`auto` -> `RetainPtr`

cdumez replied:
You can actually use `toNSData(bookmarkData)` from SpanCocoa.h

cdumez commented on Source/WebKit/NetworkProcess/Downloads/Download.h:
```diff
@@ -86,6 +86,10 @@ class Download : public IPC::MessageSender, public CanMakeWeakPtr<Download> {
 #endif
 
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    void publishProgress(const URL&, std::span<const uint8_t>&&);
```
You don't need to pass span by r-value references and use WTFMove(). spans are very cheap to copy.

cdumez commented on Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:
```diff
@@ -89,8 +93,12 @@ class PendingDownload : public NetworkLoadClient, public IPC::MessageSender, pub
 #if PLATFORM(COCOA)
     URL m_progressURL;
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    std::span<const uint8_t> m_bookmarkData;
```
Maybe this should be a Vector?

cdumez commented on Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
```diff
@@ -126,13 +126,23 @@ void DownloadManager::cancelDownload(DownloadID downloadID, CompletionHandler<vo
+void DownloadManager::publishDownloadProgress(DownloadID downloadID, const URL& url, std::span<const uint8_t> bookmarkData)
+{
+    if (auto* download = m_downloads.get(downloadID))
+        download->publishProgress(url, WTFMove(bookmarkData));
```
WTFMove() does nothing here.

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -2149,11 +2149,18 @@ void NetworkProcess::cancelDownload(DownloadID downloadID, CompletionHandler<voi
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+void NetworkProcess::publishDownloadProgress(DownloadID downloadID, const URL& url, std::span<const uint8_t> bookmarkData)
+{
+    downloadManager().publishDownloadProgress(downloadID, url, WTFMove(bookmarkData));
```
WTFMove() is not needed.

cdumez commented on Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
```diff
@@ -75,17 +75,31 @@ void PendingDownload::cancel(CompletionHandler<void(std::span<const uint8_t>)>&&
+    if (!m_progressURL.isValid())
+        return;
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    download->publishProgress(m_progressURL, WTFMove(m_bookmarkData));
```
WTFMove() seems to do nothing here?

cdumez reviewed (approved):
LGTM with nits.

pvollan committed:
Report network download progress
rdar://129296777
https://bugs.webkit.org/show_bug.cgi?id=275176

Reviewed by Chris Dumez.

Adopt modern download progress reporting mechanism. This is replacing our current use of the
NSProgress API for reporting download progress. The feature flag is not being set in this
patch, so there should be no behavior change from this patch. This patch is also replacing
the use of sandbox SPI to create and consume a sandbox extension to the location of the
downloaded file. This is being replaced with the bookmark API.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebKit/NetworkProcess/Downloads/Download.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
(WebKit::DownloadManager::publishDownloadProgress):
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
(WebKit::PendingDownload::publishProgress):
(WebKit::PendingDownload::didBecomeDownload):
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):
* Source/WebKit/NetworkProcess/Downloads/cocoa/WKDownloadProgress.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/WKDownloadProgress.mm:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::publishDownloadProgress):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
(WebKit::DownloadProxy::publishProgress): Deleted.
* Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm: Added.
(WebKit::DownloadProxy::publishProgress):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/281152@main
